src/m3ninx/index/segment/builder/multi_segments_multi_key_postings_list_iter.go

```diff
@@ -123,49 +117,74 @@ func (i *multiKeyPostingsListIterator) Next() bool {
 		return false
 	}
 
-	prevField := i.currIters[0].Current()
-
 	// Re-evaluate current value
 	i.currEvaluate()
 
 	// NB(bodu): Build the postings list for this field if the field has changed.
 	defer func() {
-		for idx := range i.currFieldPostingsLists {
-			i.currFieldPostingsLists[idx] = nil
-		}
-		i.currFieldPostingsLists = i.currFieldPostingsLists[:0]
 		for idx := range i.currReaders {
 			i.currReaders[idx] = nil
 		}
 		i.currReaders = i.currReaders[:0]
 	}()
+	i.currFieldPostingsList.Reset()
 	currField := i.currIters[0].Current()
+	for _, iter := range i.currIters {
+		fieldsKeyIter := iter.(*fieldsKeyIter)
+		reader, err := fieldsKeyIter.segment.segment.Reader()
+		if err != nil {
+			i.err = err
+			return false
+		}
+		pl, err := reader.MatchField(currField)
+		if err != nil {
+			i.err = err
+			return false
+		}
 
-	if !bytes.Equal(prevField, currField) {
-		i.currFieldPostingsList.Reset()
-		for _, segment := range i.segments {
-			reader, err := segment.Reader()
-			if err != nil {
-				i.err = err
-				return false
+		if fieldsKeyIter.segment.offset == 0 {
+			// No offset, which means is first segment we are combining from
+			// so can just direct union
+			i.currFieldPostingsList.Union(pl)
+			continue
+		}
+
+		// We have to taken into account the offset and duplicates
+		var (
+			iter           = i.bitmapIter
+			duplicates     = fieldsKeyIter.segment.duplicatesAsc
+			negativeOffset postings.ID
+		)
+		bitmap, ok := roaring.BitmapFromPostingsList(pl)
+		if !ok {
+			i.err = errPostingsListNotRoaring
+			return false
+		}
+
+		iter.Reset(bitmap)
+		for v, eof := iter.Next(); !eof; v, eof = iter.Next() {
+			curr := postings.ID(v)
+			for len(duplicates) > 0 && curr > duplicates[0] {
+				duplicates = duplicates[1:]
+				negativeOffset++
 			}
-			pl, err := reader.MatchField(currField)
-			if err != nil {
-				i.err = err
-				return false
+			if len(duplicates) > 0 && curr == duplicates[0] {
+				duplicates = duplicates[1:]
+				negativeOffset++
+				// Also skip this value, as itself is a duplicate
+				continue
 			}
-
-			i.currFieldPostingsLists = append(i.currFieldPostingsLists, pl)
-			i.currReaders = append(i.currReaders, reader)
+			value := curr + fieldsKeyIter.segment.offset - negativeOffset
+			_ = i.currFieldPostingsList.Insert(value)
 		}
 
-		i.currFieldPostingsList.UnionMany(i.currFieldPostingsLists)
+		i.currReaders = append(i.currReaders, reader)
```

Comment:
Perhaps add the reader right at the start in case you do a `continue`?  It seems like if it's the first one you do a direct union and miss adding the reader to the `i.currReaders`?